From: Samuel Angebault <staphylo@arista.com>
Date: Tue, 9 Jun 2020 14:43:31 -0700
Subject: Add loopback from file support

By providing the extra loopoffset= parameter, it becomes possible to
mount a rootfs from within the file pointed by loop= at a given offset.
This mechanism uses losetup to create a loopdevice

Refreshed for Ubuntu initramfs-tools 0.151ubuntu1: the offset now has to be
honoured both by mount_loop_root() (used by the NFS path) and by the loop
mount that Ubuntu inlined into scripts/local.

---
 init              |  5 +++++
 initramfs-tools.7 |  5 +++++
 scripts/functions |  8 ++++++++
 scripts/local     | 14 +++++++++++---
 4 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/init b/init
--- a/init
+++ b/init
@@ -63,6 +63,7 @@
 export LOOP=
 export LOOPFLAGS=
 export LOOPFSTYPE=
+export LOOPOFFSET=
 export IP=
 export IP6=
 export VLAN=
@@ -149,6 +150,10 @@
 		# shellcheck disable=SC2034
 		LOOPFSTYPE="${x#loopfstype=}"
 		;;
+	loopoffset=*)
+		# shellcheck disable=SC2034
+		LOOPOFFSET="${x#loopoffset=}"
+		;;
 	nfsroot=*)
 		# shellcheck disable=SC2034
 		NFSROOT="${x#nfsroot=}"
diff --git a/initramfs-tools.7 b/initramfs-tools.7
--- a/initramfs-tools.7
+++ b/initramfs-tools.7
@@ -71,6 +71,11 @@
 set the loop file system mount option string, if applicable.
 
 .TP
+\fB\fI loopoffset
+set the loop file offset from which to mount the loop, if applicable.
+The default is 0 and requires loop to be defined.
+
+.TP
 \fB\fI nfsroot
 can be either "auto" to try to get the relevant information from DHCP or a
 string of the form NFSSERVER:NFSPATH or NFSSERVER:NFSPATH:NFSOPTS.
diff --git a/scripts/functions b/scripts/functions
--- a/scripts/functions
+++ b/scripts/functions
@@ -1071,6 +1071,14 @@
 	modprobe loop
 	modprobe "${FSTYPE}"
 
+	if [ -n "${LOOPOFFSET}" ]; then
+		# create a loop device for the fs within the file
+		loopdev="$(losetup -f)"
+		losetup -o "${LOOPOFFSET:-0}" "${loopdev}" "${loopfile}" || \
+		    panic "ALERT! $loopdev could not be setup using $loopfile"
+		loopfile="$loopdev"
+	fi
+
 	# FIXME This has no error checking
 	if [ -z "${LOOPFLAGS}" ]; then
 		mount ${roflag} -o loop -t "${FSTYPE}" "$loopfile" "${rootmnt}"
diff --git a/scripts/local b/scripts/local
--- a/scripts/local
+++ b/scripts/local
@@ -258,9 +258,21 @@
 		# FIXME This has no error checking
 		modprobe loop
 
+		if [ -n "${LOOPOFFSET}" ]; then
+			# create a loop device for the fs within the file
+			loopdev="$(losetup -f)"
+			losetup -o "${LOOPOFFSET:-0}" "${loopdev}" "/host/${LOOP#/}" || \
+			    panic "ALERT! $loopdev could not be setup using /host/${LOOP#/}"
+			LOOPFILE_IS_DEV=y
+		fi
+
 		# FIXME This has no error checking
 		# shellcheck disable=SC2086
-		mount ${roflag} -o loop -t ${FSTYPE} ${LOOPFLAGS} "/host/${LOOP#/}" "${rootmnt?}"
+		if [ "${LOOPFILE_IS_DEV}" = "y" ]; then
+			mount ${roflag} -t ${FSTYPE} ${LOOPFLAGS} "${loopdev}" "${rootmnt?}"
+		else
+			mount ${roflag} -o loop -t ${FSTYPE} ${LOOPFLAGS} "/host/${LOOP#/}" "${rootmnt?}"
+		fi
 
 		if [ -d "$rootmnt/host" ]; then
 			mount -o move /host "$rootmnt/host"
-- 
2.26.2

